src/instrumentation.rs

```diff
@@ -0,0 +1,342 @@
+use async_trait::async_trait;
+use dashmap::DashMap;
+use dipstick::{Input, InputScope, Prefixed, Statsd, StatsdScope};
+use log::{debug, error, info, warn};
+use rdkafka::message::{BorrowedMessage, Message};
+use std::convert::TryInto;
+use std::sync::Arc;
+use std::time::Instant;
+use tokio::{
+    sync::mpsc::{channel, Receiver, Sender},
+    task,
+};
+
+pub type Statistic = (StatTypes, i64);
+
+pub fn init_stats(endpoint: &str, app_id: &str) -> Result<Arc<Sender<Statistic>>, std::io::Error> {
+    let scope = Statsd::send_to(endpoint)?.named(app_id).metrics();
+    let scope_ref = Arc::new(scope);
+
+    let mut handler = StatsHandler::new(scope_ref.clone());
+
+    let channel = Arc::new(handler.tx.clone());
+
+    task::spawn(async move {
+        handler.run_loop().await;
+    });
+
+    Ok(channel.clone())
+}
+
+#[async_trait]
+pub trait Instrumentation {
+    // messages
+
+    async fn log_message_received(&self, m: &BorrowedMessage) {
+        debug_message("Message received", m);
+        self.record_stat((StatTypes::MessageReceived, 1)).await;
+    }
+
+    async fn log_message_deserialized(&self, m: &BorrowedMessage) {
+        debug_message("Message deserialized", m);
+        self.record_stat((StatTypes::MessageDeserialized, 1)).await;
+    }
+
+    async fn log_message_deserialization_failed(&self, m: &BorrowedMessage) {
+        debug_message("Message deserialization failed", m);
+        self.record_stat((StatTypes::MessageDeserializationFailed, 1))
+            .await;
+    }
+
+    async fn log_message_transformed(&self, m: &BorrowedMessage) {
+        debug_message("Message transformed", m);
+        self.record_stat((StatTypes::MessageTransformed, 1)).await;
+    }
+
+    async fn log_message_transform_failed(&self, m: &BorrowedMessage) {
+        debug_message("Message transformed failed", m);
+        self.record_stat((StatTypes::MessageTransformFailed, 1))
+            .await;
+    }
+
+    async fn log_message_buffered(&self, m: &BorrowedMessage, message_buffer_len: usize) {
+        debug_message("Message buffered", m);
+        self.record_stat((StatTypes::MessageBuffered, 1)).await;
+        self.record_stat((StatTypes::BufferedMessages, message_buffer_len as i64))
+            .await;
+    }
+
+    // assignments
+
+    async fn log_assignment_untracked_skipped(&self, m: &BorrowedMessage) {
+        info_message("Partition is not tracked but a delta tx exists. Updating partition assignment and skipping message. Will process the same message again after consumer seek", m);
```

Comment:
You're right, it doesn't record any stats (and there's a couple others like that as well). I liked the "fit" of keeping them in this `Instrumentation` trait though so that logging/metrics related changes can be managed in one spot instead of spread out in the runloop. Maybe I'm being silly 🤔.